metacat-common-server/src/main/java/com/netflix/metacat/common/server/connectors/ValidatingConnectorCatalogService.java

```diff
@@ -18,73 +22,119 @@
 import java.util.List;
 
 /**
- * Connector that throttles calls to the connector based on the contextual request name
- * and the resource. Not all APIs can be throttles since we may not have a resource
- * but those are a small monitory
+ * Connector that validates requests based on the request context and/or resource.
  */
 @Slf4j
 @RequiredArgsConstructor
-public class ThrottlingConnectorCatalogService implements ConnectorCatalogService {
+public class ValidatingConnectorCatalogService implements ConnectorCatalogService {
     @Getter
     @NonNull
     private final ConnectorCatalogService delegate;
     @NonNull
     private final RateLimiter rateLimiter;
+    private final boolean rateLimiterEnabled;
+    @NonNull
+    private final Authorization authorization;
+    private final boolean authorizationEnabled;
 
     @Override
     public void create(final ConnectorRequestContext context, final CatalogInfo resource) {
-        checkThrottling(MetacatContextManager.getContext().getRequestName(), resource.getName());
+        if (rateLimiterEnabled) {
```

Comment:
For option2, maybe I miss something, but I feel like we don't need to have a config for each service.

I was thinking we still keep the two original config extempted setting (one for throttling and one for authorization), and have just 1 Validator interface and 2 Implementations (one for throttling and another for Authorization).

The interface can have one function say validate, which takes in MetacatRequestContext, and have one variable call enabled.

Then for the implementation for either throttling and Authorization, the implementation for validate should be if its enabled, then trigger the corresponding validation.

Then for each service (table/partition/catalog/database), it should have a list of Validator. I see most of the functions check both Throttling and Authorization, so we can have a generic util function that run all validators. But for some cases, where not all validators should run, we can handle them separately.

Reply:
i see what you mean-- this is cleaner and more extensible. we will need to be careful to include the following:
1. special handling for specific methods that do not check throttling even when enabled at the connector level
2. ensure we are explicit about the order in which different validations occur. for example, always check throttling first, then auth
3. (less important) retain the ability to switch the order of each validation at the method level. this is not needed now but could be as we add more to it

will wait for others opinions before making the change

Reply:
I prefer option 2. Not sure if Yingjian already mentioned, but we can have a Validator interface, which validates given the context. There would be 3 implementations - ThrottlingValidator, AuthorizationValidator and a ChainedValidator which takes an ordered list of Validators and executes them.  ValidatingConnectorCatalogService would take the Validator interface, but the constructor would pass in the ChainedValidator which is opinionated about the order of execution.

Reply:
Here's a suggestion which iterates over the above suggestions:

1) As Yingjian suggested, we should have a Validator interface with two implementations—ThrottlingValidator and AuthorizationValidator—that perform validations.
2) Have a ValidationRegistry class that contains all the validators registered with it. It contains a method run that takes in a list of validations to be run. This will be used to execute the validations in the order or exact order they are to be run.
3) Implement a ValidatingConnectorBase which implements a generic ConnectorBaseService. It takes the validation registry and a delegate and implements all the methods. Each method would first perform the validation via registry.
4) Similarly, implement all the specific implementations for table, partition etc; with a similar logic.